File: 99-RecoverBinarySearchTree/99-RecoverBinarySearchTree.py
# Last updated: 4/21/2026, 10:30:31 AM
1# Definition for a binary tree node.
2# class TreeNode:
3#     def __init__(self, val=0, left=None, right=None):
4#         self.val = val
5#         self.left = left
6#         self.right = right
7class Solution:
8    def recoverTree(self, root: Optional[TreeNode]) -> None:
9        """
10        Do not return anything, modify root in-place instead.
11        """
12        arr = []
13        def dfs(curr):
14            if not curr:
15                return
16            dfs(curr.left)
17            arr.append(curr)
18            dfs(curr.right)
19        dfs(root)
20        first, second = None, None
21        for i in range(1, len(arr)):
22            if arr[i - 1].val > arr[i].val:
23                if first is None:
24                    first = arr[i - 1]
25                second = arr[i]
26        first.val, second.val = second.val, first.val

File: 99-RecoverBinarySearchTree/99-RecoverBinarySearchTree.cpp
// Last updated: 9/11/2026, 5:41:43 PM
1/**
2 * Definition for a binary tree node.
3 * struct TreeNode {
4 *     int val;
5 *     TreeNode *left;
6 *     TreeNode *right;
7 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
8 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
9 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
10 * };
11 */
12class Solution {
13public:
14    void inorder(TreeNode* curr, TreeNode*& prev, TreeNode*& first, TreeNode*& second) {
15        if (!curr) return;
16        inorder(curr->left, prev, first, second);
17        if (prev && prev->val >= curr->val) {
18            if (!first) first = prev;
19            second = curr;
20        }
21        prev = curr;
22        inorder(curr->right, prev, first, second);
23    }
24    void recoverTree(TreeNode* root) {
25        TreeNode* first = nullptr;
26        TreeNode* second = nullptr;
27        TreeNode* prev = nullptr;
28        inorder(root, prev, first, second);
29        swap(first->val, second->val);
30    }
31};